include/IECoreScene/PrimitiveVariableAlgos.h

```diff
@@ -0,0 +1,299 @@
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef IECORESCENE_PRIMITIVEVARIABLEALGOS_H
+#define IECORESCENE_PRIMITIVEVARIABLEALGOS_H
+
+#include "IECore/VectorTypedData.h"
+
+#include "IECoreScene/PrimitiveVariable.h"
+
+#include <unordered_map>
+
+namespace IECoreScene
+{
+
+namespace PrimitiveVariableAlgos
+{
+
+struct IndexedData
+{
+	IndexedData()
+	{
+	}
+
+	IndexedData( IECore::DataPtr data, IECore::IntVectorDataPtr indices ) : data( data ), indices( indices )
+	{
+	}
+
+	IECore::DataPtr data;
+	IECore::IntVectorDataPtr indices;
+};
+
+/// For building a filtered indexed data array from an existing data array
+/// ensuring the resulting data array is compact.
+template<typename T, template<typename> class V>
+class IndexedPrimitiveVariableBuilder
+{
+	public:
+		IndexedPrimitiveVariableBuilder( int dataReserveSize, int indexReserveSize )
+			: m_data( new V<std::vector<T> >() ),
+			m_writable( m_data->writable() ),
+			m_indices( new IECore::IntVectorData() ),
+			m_writableIndices( m_indices->writable() )
+		{
+			m_writable.reserve( dataReserveSize );
+
+			if( indexReserveSize > 0 )
+			{
+				m_writableIndices.reserve( indexReserveSize );
+			}
+		}
+
+		void addIndexedValue( const PrimitiveVariable::IndexedView<T> &indexedData, int i )
+		{
+			if( !indexedData.indices() )
+			{
+				m_writable.push_back( indexedData[i] );
+				return;
+			}
+
+			int oldIndex = indexedData.index( i );
+
+			auto it = m_indexMapping.find( oldIndex );
+
+			if( it != m_indexMapping.end() )
+			{
+				m_writableIndices.push_back( it->second );
+			}
+			else
+			{
+				int newIndex = static_cast<int> ( m_writable.size() );
+				m_writableIndices.push_back( newIndex );
+				m_indexMapping[oldIndex] = newIndex;
+				m_writable.push_back( indexedData[i] );
+			}
+		}
+
+		IndexedData indexedData()
+		{
+			if( m_writableIndices.size() )
+			{
+				return IndexedData( m_data, m_indices );
+			}
+
+			return IndexedData( m_data, nullptr );
+
+		}
+
+	private:
+
+		typename V<std::vector<T>>::Ptr m_data;
+		std::vector<T> &m_writable;
+
+		IECore::IntVectorDataPtr m_indices;
+		std::vector<int> &m_writableIndices;
+
+		std::unordered_map<int, int> m_indexMapping;
+};
+
+// Base type for all Functors which delete primivars
+template<typename U, typename I>
```

Comment:
What's the reason for the `I` template parameter?

Reply:
I'll remove this as I was thinking we should support other index types like u16 in the future.